File: ANE2/backend-socket/test-socket/main.py
import socket
import os
import json
import time

# A simple logger-like function for demonstration
def logger(level, message):
    print(f"[{level.upper()}] {message}")

def clean_socket(socket_path):
    """
    Ensures a socket file at the given path is removed if it exists.
    This prevents "Address already in use" errors on startup.
    """
    if os.path.exists(socket_path):
        try:
            os.remove(socket_path)
            logger("info", f"Removed stale socket file: {socket_path}")
        except OSError as e:
            logger("error", f"Error removing socket file {socket_path}: {e}")

def create_socket(socket_path):
    """
    Creates and binds a UNIX domain socket.

    @param socket_path The file path to bind the socket to.
    @return The created and bound socket object.
    @raises Exception if socket creation or binding fails.
    """
    try:
        server_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        logger("info", f"Binding socket to {socket_path}...")
        server_socket.bind(socket_path)
        logger("info", "Socket bound successfully.")
        return server_socket
    except Exception as e:
        logger("error", f"Failed to create or bind socket at {socket_path}: {e}")
        raise

def main():
    socket_path = "/tmp/my_unix_socket"
    clean_socket(socket_path)

    try:
        server_socket = create_socket(socket_path)
        server_socket.listen(1)

        # This is the main server superloop
        while True:
            logger("info", "Waiting for a client to connect...")
            
            # accept() blocks here until a client connects.
            # The 'if/else' is not needed.
            connection, _ = server_socket.accept()
            
            try:
                logger("info", "Client connected.")

                # 1. Define and send the job to the client
                job_to_send = {"service": "acquire", "timeService": "5seconds"}
                message = json.dumps(job_to_send).encode('utf-8')

                logger("info", f"Sending job to client: {job_to_send}")
                connection.sendall(message)

                # 2. Wait for the processed data from the client
                logger("info", "Waiting for processed data from client...")
                data_received = connection.recv(1024)

                if data_received:
                    try:
                        decoded_data = data_received.decode('utf-8')
                        result_data = json.loads(decoded_data)
                        logger("info", f"Received result from client: {result_data}")

                        # 3. Master Logic: Validate the result from the slave
                        original_service = result_data.get("original_service")
                        result_code = result_data.get("result_code")

                        if result_code == 0 and original_service == job_to_send["service"]:
                            time_taken = result_data.get("time_taken", "N/A")
                            logger("info", f"SUCCESS: Client completed job '{original_service}' in {time_taken} seconds.")
                        else:
                            error_message = result_data.get("message", "No details provided.")
                            logger("error", f"FAILURE: Client reported an error for job '{original_service}'. Details: {error_message}")

                    except (json.JSONDecodeError, UnicodeDecodeError) as e:
                        logger("error", f"Could not decode or parse client's response: {e}")
                else:
                    # This happens if the client connects and immediately disconnects without sending data.
                    logger("error", "Connection established, but no data received from client.")
            
            except socket.error as e:
                # This can catch errors during sendall() or recv()
                logger("error", f"A socket error occurred during communication: {e}")

            finally:
                # This block ensures the connection is always closed,
                # preparing the server for the next client.
                logger("info", "Closing client connection.")
                connection.close()
                time.sleep(1) # Optional small delay

    except Exception as e:
        logger("error", f"A critical server error occurred: {e}")
    finally:
        clean_socket(socket_path)

if __name__ == "__main__":
    main()

File: ANE2/backend-socket/test-socket/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>
#include "cJSON.h" // Make sure you have the cJSON library

#define SOCKET_PATH "/tmp/my_unix_socket"
#define BUFFER_SIZE 1024

// Function to build and send the final processed data
void send_job_result(int socket, const char *service_name, int time, int code, const char *message) {
    cJSON *result_json = cJSON_CreateObject();
    if (result_json == NULL) {
        fprintf(stderr, "Failed to create JSON object for result.\n");
        return;
    }

    cJSON_AddStringToObject(result_json, "original_service", service_name);
    cJSON_AddNumberToObject(result_json, "time_taken", time);
    cJSON_AddNumberToObject(result_json, "result_code", code);
    cJSON_AddStringToObject(result_json, "message", message);

    char *result_string = cJSON_PrintUnformatted(result_json);
    if (result_string == NULL) {
        fprintf(stderr, "Failed to print result JSON to string.\n");
    } else {
        printf("Sending processed data: %s\n", result_string);
        if (send(socket, result_string, strlen(result_string), 0) == -1) {
            perror("send result error");
        }
        free(result_string); // Must free the string allocated by cJSON
    }
    cJSON_Delete(result_json); // Free the cJSON object
}


int main() {
    int client_socket;
    struct sockaddr_un server_addr;
    char buffer[BUFFER_SIZE];

    // Superloop for the client
    while (1) {
        if ((client_socket = socket(AF_UNIX, SOCK_STREAM, 0)) == -1) {
            perror("socket error");
            sleep(2);
            continue;
        }

        memset(&server_addr, 0, sizeof(struct sockaddr_un));
        server_addr.sun_family = AF_UNIX;
        strncpy(server_addr.sun_path, SOCKET_PATH, sizeof(server_addr.sun_path) - 1);

        printf("Connecting to master server...\n");
        if (connect(client_socket, (struct sockaddr *)&server_addr, sizeof(struct sockaddr_un)) == -1) {
            perror("connect error");
            close(client_socket);
            sleep(2);
            continue;
        }

        printf("Connected. Waiting for a job...\n");

        // 1. Receive the job from the server
        ssize_t num_bytes = recv(client_socket, buffer, BUFFER_SIZE - 1, 0);
        if (num_bytes > 0) {
            buffer[num_bytes] = '\0';
            printf("Received job: %s\n", buffer);

            // 2. Process the job
            cJSON *json = cJSON_Parse(buffer);
            if (json == NULL) {
                fprintf(stderr, "Error: Could not parse job JSON.\n");
                // Report the failure back to the master
                send_job_result(client_socket, "unknown", 0, -1, "Failed to parse incoming JSON job");
            } else {
                cJSON *service = cJSON_GetObjectItemCaseSensitive(json, "service");
                cJSON *timeService = cJSON_GetObjectItemCaseSensitive(json, "timeService");
                int delay_seconds = 0;
                const char *service_name_str = "unknown";

                if (cJSON_IsString(service) && (service->valuestring != NULL)) {
                    service_name_str = service->valuestring;
                    printf("Processing service: %s\n", service_name_str);
                }
                if (cJSON_IsString(timeService) && (timeService->valuestring != NULL)) {
                    // Extract integer from string like "5seconds"
                    sscanf(timeService->valuestring, "%d", &delay_seconds);
                }

                // 3. Simulate doing the work
                printf("Simulating work for %d seconds...\n", delay_seconds);
                sleep(delay_seconds > 0 ? delay_seconds : 1);
                printf("Work finished.\n");
                
                // 4. Send the "processed dummy data" back as a result
                send_job_result(client_socket, service_name_str, delay_seconds, 0, "Task completed successfully");
                
                cJSON_Delete(json);
            }
        } else {
            printf("Connection closed by server or recv error.\n");
        }

        close(client_socket);
        printf("Disconnected. Will try to reconnect.\n\n");
        sleep(2);
    }

    return 0;
}